### PR TITLE
Grib dep2

### PR DESCRIPTION
**Pull request by marqh**
deprecations and publications for the GRIB loading, to prepare for strict_grib_load to become the default and to support as_cubes (#1941)

### CODE REVIEW

File: lib/iris/fileformats/grib/message.py [hunk 1]
```diff
@@ -323,16 +323,16 @@ def _get_message_sections(self):
...
                 section_keys = []
                 section = new_section
             section_keys.append(key_name)
-        sections[section] = _Section(self._message_id, section, section_keys)
+        sections[section] = Section(self._message_id, section, section_keys)
         return sections
 
 
-class _Section(object):
+class Section(object):
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: Needs a docstring

**Comment by ajdawson** [comment]
Can you comment in some more detail why you want to make a large chunk of the deliberately private grib API public? This seems like an unusual step considering the difficulties we have had already with modifying public APIs (complex deprecation cycles and waiting for major releases).

**Comment by marqh** [comment]
> Can you comment in some more detail why you want to make a large chunk of the deliberately private grib API public? This seems like an unusual step considering the difficulties we have had already with modifying public APIs (complex deprecation cycles and waiting for major releases).

of course

this is not without some consideration and a measure of trepidation

#1941 provides an important functionality extension, required by a number of users: it enables the use of 'file format fields' such as GRIB messages as feeds into the convert process.  Currently only file names are well supported

For this to be useful, we have to interact with messages.  We can do that with the legacy GRIB interface, but this PR is deprecating that, to enable deletion at a future release.

To deprecate these and support #1941 we need to be able to work with GRIB messages, and all the newer code is private.

So, I have promoted 2 Classes to the public API to enable us to do this

**Comment by ajdawson** [comment]
Wehn you say it needs to be public, does it need to be visible to the end-user in order for the pairs loading to work, or just used by those modules?

**Comment by marqh** [comment]
> When you say it needs to be public, does it need to be visible to the end-user in order for the pairs loading to work, or just used by those modules?

this needs to be visible to the end user in order for the pairs loading to be useful for users.

**Comment by pp-mo** [comment]
> @marqh this needs to be visible to the end user in order for the pairs loading to be useful for users

I'd support that:
By  [#1941 "as_cubes"](#1941), we now have new routines (```messages_from_filename``` and ```load_pairs_from_fields```) which handle _GribMessage objects.
The whole point of that is so that you can do stuff with them (if only to filter on their contents).
So their definition must be public.

**Comment by pp-mo** [comment]
As a general point, I think there should be more pointers here to go with the statements "X is deprecated" and add "go look at Y instead".
I'll add some short notes in a few places in the code...

File: docs/iris/src/whatsnew/contributions_1.10/deprecate_2016-Mar-17_GribWrapper.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+* deprecated:
+ iris.fileformats.grib.hindcast_workaround,
+ :class: `iris.fileformats.grib.GribWrapper`.
```
Comment (pp-mo) [inline docs/iris/src/whatsnew/contributions_1.10/deprecate_2016-Mar-17_GribWrapper.txt]: cf : [comment: add links into deprecation notices](https://github.com/SciTools/iris/pull/1962#issuecomment-197895178)
Reference ```GribMessage`` as an alternative.

File: lib/iris/fileformats/grib/__init__.py [hunk 3]
```diff
@@ -175,8 +176,12 @@ class GribWrapper(object):
     """
     Contains a pygrib object plus some extra keys of our own.
 
+    .. deprecated:: 1.10
+
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/__init__.py]: cf : [comment: add links into deprecation notices](https://github.com/SciTools/iris/pull/1962#issuecomment-197895178)
Reference ```GribMessage`` as an alternative.

File: lib/iris/fileformats/grib/__init__.py [hunk 4]
```diff
@@ -175,8 +176,12 @@ class GribWrapper(object):
...
+
     """
     def __init__(self, grib_message, grib_fh=None, auto_regularise=True):
+        warnings.warn('Deprecated at version 1.10')
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/__init__.py]: cf : [comment: add links into deprecation notices](https://github.com/SciTools/iris/pull/1962#issuecomment-197895178)
Maybe here too?

File: lib/iris/fileformats/grib/__init__.py [hunk 5]
```diff
@@ -834,6 +839,8 @@ def grib_generator(filename, auto_regularise=True):
     Returns a generator of :class:`~iris.fileformats.grib.GribWrapper`
     fields from the given filename.
 
+    .. deprecated:: 1.10
+
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/__init__.py]: cf : [comment: add links into deprecation notices](https://github.com/SciTools/iris/pull/1962#issuecomment-197895178)
Reference ```fileformats.message.GribMessage.messages_from_filename`` as an alternative.

File: lib/iris/fileformats/grib/__init__.py [hunk 6]
```diff
@@ -962,7 +970,7 @@ def as_pairs(cube):
 def as_messages(cube):
     """
     Convert one or more cubes to GRIB messages.
-    Returns an iterable of grib_api GRIB messages.
+    Returns an iterable of grib_api GRIB message IDs.
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/__init__.py]: I think 'handles' would be lot clearer than 'IDs', as Python objects have 'ids'.  But then I see that 'id' is used in the grib_api documentation.

This applies to all the last 3 low-level routines : as_pairs / as_messages / save_messages.
At least this one is explicit in referring to **_"grib_api_** GRIB message IDs"
I can't create a comment for it, but I think the line 950 docstring should be improved to a similar style.

Comment (pp-mo) [inline lib/iris/fileformats/grib/__init__.py]: Actually, there is some serious potential for confusion here, when we make 'message.GribMessage' public, about the fact that 'as_pairs' and 'as_messages' do **_not_** handle those, but grib_api handles instead.
I.E. despite losing 'GribWrapper', we still have a low-level and a higher-level message concept.

We could consider renaming 'GribMessage' to something that made it clearer that this is **_not_** the things that the grib_api deals in ?  I can now see that was the point of the name 'GribWrapper' !

File: lib/iris/fileformats/grib/message.py [hunk 7]
```diff
@@ -60,7 +60,7 @@ def messages_from_filename(filename):
                     break
                 raw_message = _RawGribMessage(grib_id)
                 recreate_raw = _MessageLocation(filename, offset)
-                yield _GribMessage(raw_message, recreate_raw)
+                yield GribMessage(raw_message, recreate_raw)
 
     def __init__(self, raw_message, recreate_raw):
         """
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: The docs don't explain the meaning of 'recreate_raw' at all.
I think that needs fixing now we're making it public.

Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: Also needed (but outside the scope of the current diffs):
The 'GribMessage.sections' property should have a docstring explaining that it is a **_mapping_**
 i.e.  " int ::--> :class:`Section` "
A quick usage example should be worth a lot of prose there, I think.

File: lib/iris/fileformats/grib/message.py [hunk 8]
```diff
@@ -33,7 +33,7 @@
 from iris.exceptions import TranslationError
 
 
-class _GribMessage(object):
+class GribMessage(object):
     """
     Lightweight GRIB message wrapper, containing **only** the coded keys and
     data attribute of the input GRIB message.
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: This dosctring is basically a clone of the old 'GribWrapper' one, and wants revisiting.
I don't know if we still have an ambition to limit it to coded keys, but at present of course it does not so maybe best not even to say that.
Suggest something like.. """An object wrapping a grib_api message ID and exposing grib keys via the "sections" property."""

**Comment by pp-mo** [comment]
After a longer think + further trawl through the code, it seems to me that the existing "GribMessage" class is not really ready yet to be exposed as a public general-purpose solution.  It seems at present to be designed for only specific, limited purposes : in particular, as required by the **_loading_** operation -- any possibility of using it for saving as well will require a good deal more work.
A number of points emerged ...
 * you can't modify keys in a Section, which limits the potential for customising the loading process (let alone saving)
 * it makes no good sense for a user to create a GribMessage via the normal \_\_init__ constructor --especially as, I think, we'd still much rather keep _RawGribMessage and  _MessageLocation private :  So, you can really only get valid ones from ```GribMessage.messages_from_filename```
   * in particular, you can't sensibly create a 'new from scratch' GribMessage, independent from a file, e.g. for use in saving.
 * the proposal in #1941 adds a ```load_pairs_from_fields``` function, to return GribMessage objects.  However, the existing **_save_** functionality, and the rules operation,  uses grib_api handles, and there is currently no easy route to rationalising the 'save' side :  we can't use GribMessage in its current form.
 * the caching of keys is currently weird + clumsy, given the code in ```RawGribMessage``` :
   * the code in ```Section``` looks like it intends to caches keys independently, but ...
   * GribMessage.sections --> RawGribMessage.sections --> RawGribMessage._get_message_sections()
   * ... so in fact it will read the whole lot at once, but only at the first attempt to access anything
 * although data access can be deferred till after the source file is closed, the key access cannot : It would fail if the original grib_api handle was discarded before reading any keys.

The issue about closing files is awkward because GribMessage.messages_from_filename() will close the file when the iteration is terminated.  So fine if you deal with each message in a generator, but not if you make a list of all to process them 'later'.

**_If_** we decide we still want to publish something at least barely usable in this area, 
then I suggest the best immediate resolution we can make is to force the loading of the sections, and document the resulting message objects as **_read-only_**.
 * it's possibly a bit of a disappointment not to be able to modify the messages within the load chain, as then you can't leverage the action of the rules on a modified message.  Nor can you fix a message that triggers a rules error.
 * however, you can still inspect messages and modify cubes at the 'load_pairs_...' level.
 * ...and even more simply, you can still just filter messages before generating cubes, which can help with slow loading or bad messages

**Comment by pp-mo** [comment]
> the existing "GribMessage" class is not really ready yet

Regardless of the above...
**_If_** we decide we still want to publish something at least barely usable in this area, 
then I suggest the best immediate resolution we can make is to force the loading of the sections, and document the resulting message objects as **_read-only_**.
 * it's possibly a bit of a disappointment not to be able to modify the messages within the load chain, as then you can't leverage the action of the rules on a modified message.  Nor can you fix a message that triggers a rules error.
 * however, you can still inspect messages and modify cubes at the 'load_pairs_...' level.
 * ...and even more simply, you can still just filter messages before generating cubes, which can help with slow loading or bad messages

File: lib/iris/fileformats/grib/__init__.py (hunk 6 above)
Comment (marqh) [inline lib/iris/fileformats/grib/__init__.py]: in my mind, a key deliverable for this change is to ensure that there is no need to use the grib_api directly

So, I would like to put together a proposal which gets all the functionality we need from GribMessage

There is a planned update to as_pairs and as_messages in #1941, which we can adapt to meet this aim

http://scitools.org.uk/iris/docs/v1.9.2/userguide/saving_iris_cubes.html#customising-the-save-process
would look much better if it did not include the gribapi calls, imo

**Comment by pp-mo** [comment]
Correction:
> the caching of keys is currently weird + clumsy ...  in fact it will read the whole lot at once

Creating section does not fetch all key _values_, it only gets a list of key names.

**Comment by pp-mo** [comment]
I think latest changes are an improvement, but I'm partial !

Probably now wants some tests to exercise ```GribMessage.__setitem__``` and (possibly) ```_cache_header```.

Also, I'm still wondering if naming "GribMessage" as something else might avoid any further confusion with gribapi message handles (similarly named in various existing documentation).
But I can't think of anything much better than "GribField", which is still pretty awful.

File: lib/iris/fileformats/grib/message.py [hunk 9]
```diff
@@ -259,6 +274,16 @@ def __init__(self, message_id):
         self._message_id = message_id
         self._sections = None
 
+    def _cache_header(self):
+        """
+        Create sections and cache the header keys and values
+        as the file will be closed at some point.
+
+        """
+        for section in self.sections:
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: Needs to be ```for section in self.sections.values():```.
N.B. existing works in Python3 !

**Comment by marqh** [comment]
I have pre-cached the header, such that it is usable outside of the iterator, to see the effect

some timings for integration.test_grib2:

with this change:
    2.530 (sigma 0.173)

master:
    0.835 (sigma 0.016)

@pp-mo  as you suspected, the _cache_sections is pretty expensive, a factor 3 increase in pain

hmmm....

**Comment by pp-mo** [comment]
> a factor 3 increase in pain

Sounds big, but it's only significant in terms of some "total" operation time.
I'd say it's reasonable to measure iris.raw_load, at least, i.e. to include rules processing.
Are these numbers seconds, and for what operation over how many messages ?

File: lib/iris/fileformats/grib/message.py (hunk 9 above)
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: As it is now, the thing called 'section' is an index.
But it's only used the context "self.sections[section]".
So I still think it's clearer to iterate over the actual section objects...
```
    for section in self.sections.values():
        for key in section.keys():
            ....
```

File: lib/iris/fileformats/grib/__init__.py (hunk 6 above)
Comment (marqh) [inline lib/iris/fileformats/grib/__init__.py]: hmm, ok; this isn't going to be delivered by in this PR, we will ahve to wait until we are confident in message.GribMessage as a file writer, with appropriate grib save refactoring, before we can deliver this

File: lib/iris/fileformats/grib/message.py [hunk 10]
```diff
@@ -53,27 +62,28 @@ def messages_from_filename(filename):
...
+        file_ref = _OpenFileRef(grib_fh)
+
+        while True:
+            offset = grib_fh.tell()
+            grib_id = gribapi.grib_new_from_file(grib_fh)
+            if grib_id is None:
+                break
+            raw_message = _RawGribMessage(grib_id)
+            recreate_raw = _MessageLocation(filename, offset)
+            yield GribMessage(raw_message, recreate_raw)
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: Important bit missing : should say "GribMessage(raw_message, recreate_raw**, file_ref=file_ref**)"

File: lib/iris/fileformats/grib/message.py [hunk 11]
```diff
@@ -400,7 +391,16 @@ def __getitem__(self, key):
...
                     key, self._number))
             else:
-                value = self._get_key_value(key)
+                # _get_key_value can fail, if the gribapi has released
+                # the message.  In this case, the user can recreate the
+                # message an cache all the unread headers
+                try:
+                    value = self._get_key_value(key)
+                except gribapi.GribInternalError:
+                    msg = ('The underlying grib message is no longer'
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: Probably now obsolete ?

**Comment by marqh** [comment]
ok, this is now looking like a coherent change.

The message module, the GribMessage and the Section classes are now public.  
The message maintains a link to it's file until references to it are removed, so it can be used outside of the from_filenames yield.
The cache on each section is the primary access point.  __setitem__ updates this cache and leaves the file on disk alone

I think this represents a good access pattern for now, able to support #1941 (I hope)

@pp-mo has been instrumental in this implementation.
@rhattersley @ajdawson we would greatly value your opinions on this implementation and the API additions at this stage

File: lib/iris/fileformats/grib/message.py [hunk 12]
```diff
@@ -52,31 +66,46 @@ def messages_from_filename(filename):
...
-
-        * raw_message:
-            The _RawGribMessage instance which should be wrapped to
-            provide the `data` attribute.
+        It is recommended to obtain GribMessage instance from the static method
+        :meth:`GribMessage.messages_from_filename`, rather than creating
+        them directly.
 
         """
         self._raw_message = raw_message
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: Some (private) comments here clarifying the purpose of the constructor args would be good.
Maybe something like... :
```
    # A RawGribMessage giving gribapi access to the original grib message.
    self._raw_message = raw_message
    # A _MessageLocation which biggus uses to read the message data array,
    # by which time this message may be dead and the original grib file closed.
    self._recreate_raw = recreate_raw
    # An _OpenFileRef to keep the grib file open while this GribMessage is alive,
    # so that we can always use self._raw_message to fetch keys.
    self._file_ref = file_ref
```

I admit that seems a bit wordy though...

File: lib/iris/fileformats/grib/__init__.py [hunk 13]
```diff
@@ -55,6 +55,7 @@
 
 #: Set this flag to True to enable support of negative forecast periods
 #: when loading and saving GRIB files.
+#: .. deprecated:: 1.10
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/__init__.py]: An extra "#:" spacing line is wanted before the 'deprecated' line, to make this display nicely.

File: lib/iris/fileformats/grib/message.py [hunk 14]
```diff
@@ -33,17 +33,31 @@
...
+
+    def __del__(self):
+        if not self.open_file.closed:
+            self.open_file.close()
+
+
+class GribMessage(object):
+    """
+    An in-memory representation of a GribMessage, providing
+    access to the :meth:`GribMessage.data` payload and the metadata
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: These display as "GribMessage.data()" and "GribMessage.sections()".
You can avoid the noise of repeating the module path, by using the form :meth:`~GribMessage.data`  (i.e. add a tilde)

File: lib/iris/fileformats/grib/message.py [hunk 15]
```diff
@@ -33,17 +33,31 @@
...
+    elements by section via the :meth:`GribMessage.sections` property.
 
     """
 
     @staticmethod
     def messages_from_filename(filename):
         """
-        Return a generator of :class:`_GribMessage` instances; one for
+        Return a generator of :class:`GribMessage` instances; one for
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: "~" useful here too.

File: lib/iris/fileformats/grib/message.py [hunk 16]
```diff
@@ -52,31 +66,53 @@ def messages_from_filename(filename):
...
+    def __init__(self, raw_message, recreate_raw, file_ref=None):
         """
-
-        Args:
-
-        * raw_message:
-            The _RawGribMessage instance which should be wrapped to
-            provide the `data` attribute.
+        It is recommended to obtain GribMessage instance from the static method
+        :meth:`GribMessage.messages_from_filename`, rather than creating
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: "~" useful here too.

File: lib/iris/fileformats/grib/message.py [hunk 17]
```diff
@@ -52,31 +66,53 @@ def messages_from_filename(filename):
...
+        # alive, so that we can always use self._raw_message to fetch keys.
+        self._file_ref = file_ref
 
     @property
     def sections(self):
+        """
+        Return the key-value pairs of the message keys, grouped by containing
+        section.
+
+        Key-value pairs are collected into a dictionary of
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: Maybe doesn't read that clearly. 
"collected into" is describing from the implementation POV rather than usage.
Maybe something like this?..
```
Sections in a message are indexed by GRIB section-number, and values in a section are indexed by key strings.

.. For example::

    print(grib_message.sections[4]['parameterNumber'])
    grib_message.sections[1]['minute'] = 0
```

File: lib/iris/fileformats/grib/message.py [hunk 18]
```diff
@@ -52,31 +66,53 @@ def messages_from_filename(filename):
...
+        :class:`Section` objects. One such object is made for
+        each section in the message, such that the section number is the
+        object's key in the containing dictionary. Each object contains
+        key-value pairs for all of the message keys in the given section.
+
+        For example::
+
+            print(grib_message[4]['parameterNumber'])
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: Should read ```grib_message```**```.sections```**```[4]```
See also above note..

File: lib/iris/fileformats/grib/message.py [hunk 19]
```diff
@@ -124,6 +160,15 @@ def data(self):
             raise TranslationError(fmt.format(template))
         return data
 
+    def __getstate__(self):
```
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: I'm not convinced we need this, or for that matter any of "tests.integration.test_pickle.TestGribMessage".
As far as I can see, messages don't form any part of a cube :
only the ```fileformats.grib.message._DataProxy``` and ```._MessageLocation```do that
 * If this is ***not*** needed it should really go (and probably that pickle test too).
 * or, if it ***is*** needed then this doesn't necessarily solve the problem, as an unpickled message could not access any keys not already read in.

@rhattersley can you shed light on why we though we needed to **pickle** grib messages,
and could this maybe now be obsolete ??

**Comment by pp-mo** [comment]
@marqh sorry for pickiness.
Really close now !

I would like to pursue the mysterious pickling requirement, though.

File: lib/iris/fileformats/grib/message.py (hunk 19 above)
Comment (marqh) [inline lib/iris/fileformats/grib/message.py]: > If this is not needed it should really go (and probably that pickle test too).
> 
>     ( Else we are just passing on the burden of not knowing why this is needed ! )

May I suggest, given the lack of feedback, that this is an issue that I can take on and follow up in a subsequent PR as required, rather than blocking this issue.

> or, if it is needed then this doesn't necessarily solve the problem, as an unpickled message could not access any keys not already read in.

I disagee a little, as the unpickled current implementation has the same feature, if the keys are not cached, then they are inaccessible following an unpickle

So, may I suggest that I create and follow up on the 'why is this here, is it needed' but we don't hold off on this PR to wait for that.  This is a blocking PR already, blocking a blocking PR awaiting more input makes me feel :(

**Comment by marqh** [comment]
s/pickiness/vital feedback

suggestions adopted for docs, much clearer thank you

File: lib/iris/fileformats/grib/message.py (hunk 19 above)
Comment (pp-mo) [inline lib/iris/fileformats/grib/message.py]: Okay, on reflection I think you're right -- previous behaviour has not changed anyway, and this tweak just sidesteps the resulting exception.

**Comment by pp-mo** [comment]
Thanks @marqh for sticking with it, sorry if I dragged my feet a bit
